File: frontend/types/food.ts
export interface Ingredient {
  name: string;
  amount: string;
}

export interface MacroNutrients {
  protein_g: number | null;
  carbs_g: number | null;
  fat_g: number | null;
}

export interface FoodLog {
  id: string;
  user_id: string;
  food_name: string;
  ingredients: Ingredient[];
  serving_size: string | null;
  origin: string | null;
  calories: number;
  protein_g: number | null;
  carbs_g: number | null;
  fat_g: number | null;
  confidence: number | null;
  notes: string | null;
  image_url: string | null;
  created_at: string;
}

export interface GeminiAnalysis {
  food_name: string;
  estimated_total_calories: number;
  macros: {
    protein_g: number | null;
    carbs_g: number | null;
    fat_g: number | null;
    fibre_g: number | null;
  };
  ambiguity_flags: string[];
  overall_confidence: number;
  notes: string | null;
}

export interface AnalyzeResponse {
  analysis: GeminiAnalysis;
  inference_log_id: string;
}

export interface FoodLogResponse {
  id: string;
  user_id: string;
  meal_type: string;
  meal_name: string | null;
  estimated_total_calories: number;
  protein_g: number | null;
  carbs_g: number | null;
  fat_g: number | null;
  fibre_g: number | null;
  image_url: string[] | null;
  inference_log_id: string | null;
  created_at: string;
}

export interface FoodLogList {
  items: FoodLogResponse[];
  total: number;
}

export interface DailySummary {
  date: string;
  total_calories: number;
  total_protein_g: number;
  total_carbs_g: number;
  total_fat_g: number;
  meal_count: number;
}

export interface WeeklySummary {
  week: DailySummary[];
}

File: backend/app/schemas/food.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field


class Ingredient(BaseModel):
    name: str
    confidence: float | None = None


class Quantity(BaseModel):
    value: float
    unit: str


class MacroNutrients(BaseModel):
    protein_g: float | None = None
    carbs_g: float | None = None
    fat_g: float | None = None
    fibre_g: float | None = None
    sugar_g: float | None = None
    sodium_g: float | None = None


class GeminiAnalysis(BaseModel):
    food_name: str
    serving_size: str | None = None
    estimated_quantity: Quantity | None = None
    preparation_method: str | None = None
    ingredients: list[Ingredient] = []
    possible_alternatives: list[str] = []
    visible_sauces_or_oils: bool = False
    cuisine_type: str | None = None
    restaurant_or_brand: str | None = None
    origin: str | None = None
    estimated_total_calories: int
    macros: MacroNutrients = Field(default_factory=MacroNutrients)
    overall_confidence: float = Field(default=0.0, ge=0.0, le=1.0)
    ambiguity_flags: list[str] = []
    notes: str | None = None


class FoodLogList(BaseModel):
    items: list["FoodLogResponse"]
    total: int


class FoodLogCreate(BaseModel):
    user_id: str
    meal_type: str
    image_urls: list[str] = []
    analysis: GeminiAnalysis


class FoodLogConfirm(BaseModel):
    meal_type: str
    image_urls: list[str] = []
    analysis: GeminiAnalysis


class FoodLogUpdate(BaseModel):
    meal_type: str | None = None
    meal_name: str | None = None
    serving_size: str | None = None
    estimated_total_calories: int | None = None
    protein_g: float | None = None
    carbs_g: float | None = None
    fat_g: float | None = None
    notes: str | None = None


class FoodLogResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    user_id: str
    meal_type: str
    meal_name: str | None = None
    ingredients: list
    serving_size: str | None = None
    preparation_method: str | None = None
    estimated_total_calories: int
    protein_g: float | None = None
    carbs_g: float | None = None
    fat_g: float | None = None
    fibre_g: float | None = None
    sugar_g: float | None = None
    sodium_g: float | None = None
    visible_sauces_or_oils: bool = False
    cuisine_type: str | None = None
    restaurant_or_brand: str | None = None
    confidence: float | None = None
    ambiguity_flags: list = []
    notes: str | None = None
    image_url: list = []
    created_at: datetime

